Add toolbar keyboard shortcuts

gabrielgcma opened the pull request:
Hi folks, this is my very first community PR ever, and I'm also doing this to learn Rust, so bear that in mind.

## Description
This PR introduces changes to allow tool selection via keyboard shorcuts and closes https://github.com/gabm/Satty/issues/50.

| Tool      | Shortcut |
|-----------|----------|
| Pointer   | Alt+P    |
| Crop      | Alt+C    |
| Brush     | Alt+B    |
| Line      | Alt+L    |
| Arrow     | Alt+A    |
| Rectangle | Alt+R    |
| Ellipse   | Alt+E    |
| Text      | Alt+T    |
| Marker    | Alt+M    |
| Blur      | Alt+U    |
| Highlight | Alt+H    |

This is the event flow I came up with. The mouse click also updates the toolbar UI `active_button` so that you don't have two buttons in the UI selected at the same time (one from keyboard and another from a click).

![satty-20250102-01:36:50](https://github.com/user-attachments/assets/02994246-dbfb-4698-bec4-c2ac2919738c)

## Changes made:
Adds keyboard shortcuts functionality to the editing tools

## Notes
- I took the opportunity to refactor the input keys handling on `sketch_board.rs`.
- The `tool_shortcuts` hashmap was defined inside the `update()` method of the `SketchBoard` component for skill issue reasons. I'd like to know how to improve that.
- I'm sure there are many improvements to be made and I don't know if my approach is the best, I'm a total Rust noob. Upon testing everything seems fine, but I'd be surprised if no bugs were introduced. I'd like to know your thoughts on this.

gabm commented:
Thank you for this PR, great to see movement on this front.

I see you solved the event problem by routing. Another approach would have been a global state - but either way seems fine to me.

What I envisioned for the shortcuts would be letters and numbers without modifier keys, just as Excalidraw is doing it. That would mean, that we need to decide whether the active tool handled a certain key or not - and if not, then take action for selecting another tool. Unfortunately, at the moment we don't have a proper event handling signal for the tools. 

What should happen imho is:

- the `handle_*` functions of the Tool trait should return "handled" or "not handled", indicating whether an event was of interest to the tool or not. For example: the text tool - while in editing mode - would return handled for any key. The Arrow tool would return handled for the shift key, but not for any letter .. etc..
- the Tool trait should get a `dirty()` function. This returns a bool and indicates, whether redrawing is required
- event handling in Sketchboard is then done based on what the handle functions return. after all event handling has completed, dirty() should determine, whether redraw is now needed.

gabm commented:
Now that I think of it, another problem might be multi-language input. This will not forward single characters as "keys".

JarKz commented:
@gabm , I'm so sorry, this PR was closed by my mistake. Reopen it, please

gabrielgcma commented:
Hi guys, been busy with the new year. I intend to move forward with this PR, but I guess it will take longer than I expected. Thanks @JarKz for the addition.

@gabm 
Thanks for the reply.

> What I envisioned for the shortcuts would be letters and numbers without modifier keys, just as Excalidraw is doing it. That would mean, that we need to decide whether the active tool handled a certain key or not - and if not, then take action for selecting another tool. Unfortunately, at the moment we don't have a proper event handling signal for the tools.

Yeah, my first try included no modifiers but no success. It took me a couple hours and only now I'm starting to grasp the modifications needed to implement single-key event handling. 

> What should happen imho is:
> 
> * the `handle_*` functions of the Tool trait should return "handled" or "not handled", indicating whether an event was of interest to the tool or not. For example: the text tool - while in editing mode - would return handled for any key. The Arrow tool would return handled for the shift key, but not for any letter .. etc..
> * the Tool trait should get a `dirty()` function. This returns a bool and indicates, whether redrawing is required
> * event handling in Sketchboard is then done based on what the handle functions return. after all event handling has completed, dirty() should determine, whether redraw is now needed.

I don't quite get the difference between `dirty()` and the `handled/not handled` return type. `dirty()` would query the tool's updated `handled/not handled` and based on it return `ToolUpdateResult::Redraw`?

gabm commented:
please ignore my comment about handling and dirty functions for now (points are valid imho!), the second comment above is a "blocking" problem.

right now, all keyboard input is routed through the im multicontext for multi language input. this swallows all keys and outputs "text", which is sometimes even multiple characters and so on. With that in place, it's impossible to act on single key shortcuts. 

Further, IM multiinput has some reaction on wether a gtk widget claimed the input focus or not and handles input itself or not. 

Then there is also a gtk framework for shortcuts, which can be afaik single keys as well.

In a ideal world, we would "immitate" the widget focus behaviour to control whether im multi Input should be active or not and use the buit-in gtk shortcut mechanism for all shortcuts - including the existing ones. But to pull this off, i think we should talk to gtk devs and/or the rnote developer. Both should know how to solve this properly...

Im sorry, but I'm no expert on gtk input and the presence of multi language input (think Chinese, Japanese, etc) makes things more complicated.

gabrielgcma commented:
> please ignore my comment about handling and dirty functions for now (points are valid imho!), the second comment above is a "blocking" problem.
> 
> right now, all keyboard input is routed through the im multicontext for multi language input. this swallows all keys and outputs "text", which is sometimes even multiple characters and so on. With that in place, it's impossible to act on single key shortcuts.
> 
> Further, IM multiinput has some reaction on wether a gtk widget claimed the input focus or not and handles input itself or not.
> 
> Then there is also a gtk framework for shortcuts, which can be afaik single keys as well.
> 
> In a ideal world, we would "immitate" the widget focus behaviour to control whether im multi Input should be active or not and use the buit-in gtk shortcut mechanism for all shortcuts - including the existing ones. But to pull this off, i think we should talk to gtk devs and/or the rnote developer. Both should know how to solve this properly...
> 
> Im sorry, but I'm no expert on gtk input and the presence of multi language input (think Chinese, Japanese, etc) makes things more complicated.

Thanks for the clarifications. I think a workaround for this could be to bypass the IMMulticontext and only parse if the current tool needs it: 

```rust
                #[wrap(Some)]
                set_im_context = &gtk::IMMulticontext {
                    connect_commit[sketch_board_sender, sketchboard] => move |_cx, txt| {
                        if sketchboard.active_tool = Tools::Text {
                            sketch_board_sender.emit(SketchBoardInput::new_text_event(
                            TextEventMsg::Commit(txt.to_string()),
                        ))
                        }
                    }
                },
```

But I'm still to figure out how to satisfy the Rust compiler in order to be able access the sketchboard's state, and I don't know if skipping the code inside the if statement classifies as bypassing the IM.

gabrielgcma commented:
Hi @gabm, I think I got something, single-key shortcuts are working but I don't know if my approach is ideal. Most recent changes:

- I got single-key shortcuts working by customizing the `SketchBoard` controller's `IMContext` commit handling, and deciding wether or not to process the commited single-key press as a text event or a shortcut event by checking if the current active tool is `Tools::Text`. I'm assuming that the `IMMulticontext` is only needed for raw text inputs. If it's needed for shortcuts aswell, then my approach is flawed.

```rust
// [...]

#[wrap(Some)]
set_im_context = &gtk::IMMulticontext {
    connect_commit[sender] => move |_cx, txt| {
        sender.input(SketchBoardInput::new_commit_event(TextEventMsg::Commit(txt.to_string())));
    },
},

// [...]

fn handle_text_commit(
    &self,
    event: TextEventMsg,
    sender: ComponentSender<Self>,
) -> ToolUpdateResult {
    let tool_shortcuts = HashMap::from([
        ("p", Tools::Pointer),
        ("c", Tools::Crop),
        ("b", Tools::Brush),
        ("l", Tools::Line),
        ("a", Tools::Arrow),
        ("r", Tools::Rectangle),
        ("e", Tools::Ellipse),
        ("t", Tools::Text),
        ("m", Tools::Marker),
        ("u", Tools::Blur),
        ("h", Tools::Highlight),
    ]);
    match event {
        TextEventMsg::Commit(txt) => {
            // NOTE:
            // If there's an IMContext binded to the controller, single letter-key events will
            // always go through it first, denying a bypass, so the only way we can do single-key
            // bindings is to act upon the IMMulticontext's commit event itself.
            // NOTE:
            // Here we're basically bypassing the IMMulticontext. If the text tool is active,
            // we're interested in the single-letter keypress as a text character. If not, we
            // parse it as a shortcut event.
            if self.active_tool_type() == Tools::Text {
                sender.input(SketchBoardInput::new_text_event(TextEventMsg::Commit(
                    txt.to_string(),
                )));
                ToolUpdateResult::Unmodified
            } else {
                let key = txt.as_str();
                if let Some(tool) = tool_shortcuts
                    .iter()
                    .find(|item| *item.0 == key)
                    .map(|(_, tool)| tool)
                {
                    sender.input(SketchBoardInput::ToolbarEvent(ToolbarEvent::ToolSelected(
                        *tool,
                    )));
                    sender
                        .output_sender()
                        .emit(SketchBoardOutput::ToolSwitchShortcut(*tool));

                    if *tool == Tools::Crop {
                        ToolUpdateResult::Redraw
                    } else {
                        ToolUpdateResult::Unmodified
                    }
                } else {
                    ToolUpdateResult::Unmodified
                }
            }
        }
    }
}
```

| Tool      | Shortcut |
|-----------|----------|
| Pointer   | P    |
| Crop      | C    |
| Brush     | B    |
| Line      | L    |
| Arrow     | A    |
| Rectangle | R    |
| Ellipse   | E    |
| Text      | T    |
| Marker    | M    |
| Blur      | U    |
| Highlight | H    |

- Added `SketchBoard.get_active_tool_type()` and `Tool.get_tool_type()` for convenience.

- Moved the `gtk::EventControllerKey` declaration from the `App` to the `Sketchboard` as it was intended by a previous comment.

gabm commented:
I think that is a practical solution that is also not very intrusive. There might be problems with other keyboard layouts, but I guess it's not hurting either atm.

Until we find a "proper" solution, this will work! 

I ask you for 2 small changes for now: 

- please add a function to the tool trait that indicates, wheter a tool wants to accept input. For example, the text tool does not always process text.
- please add the numbers as well to select the tools according to their order in the toolbar...

gabrielgcma commented:
@gabm changes implemented as requested. We're checking if `TextTool` wants inputs before parsing single keys as text. This prevents `TextTools` from eating single-key presses as text when the user isn't typing.

Current shortcuts:

| Tool      | Shortcut |
|-----------|----------|
| Pointer   | P   |
| Crop      | `C`  or `1` |
| Brush     | `B` or  `2`  |
| Line      | `L` or `3`  |
| Arrow     | `A` or `4`   |
| Rectangle | `R` or `5`   |
| Ellipse   | `E` or `6`   |
| Text      | `T` or `7`   |
| Marker    | `M` or `8`    |
| Blur      | `U` or `9`    |
| Highlight | `H` or `0`    |

Let me know if something needs changing.

gabm commented:
Looks good, but there are conflicts. I suggest rebasing to latest main

gabrielgcma commented:
@gabm Had a bit of a hassle rebasing and pushing it to my own fork, I ended up deleting this PR branch. Opened another one: #148
